static/js/recent_topics.js

```diff
@@ -0,0 +1,50 @@
+const topics = new Map(); // Key is stream-id:topic.
+
+exports.process_messages = function (messages) {
+    for (const msg of messages) {
+        exports.process_message(msg);
+    }
+};
+
+exports.process_message = function (msg) {
+    if (msg.type !== 'stream') {
+        return false;
+    }
+
+    const is_ours = people.is_my_user_id(msg.sender_id);
+    const key = msg.stream_id + ':' + msg.topic;
+    const topic = topics.get(key);
+
+    if (topic === undefined) {
+        topics.set(key, {
+            last_msg_id: msg.id,
+            starred: msg.starred ? new Set([msg.id]) : new Set(),
+            participated: is_ours,
+        });
+        return true;
+    }
+
+    if (topic.last_msg_id < msg.id) {
+        topic.last_msg_id = msg.id;
+    }
+
+    if (msg.starred) {
+        topic.starred.add(msg.id);
+    }
+    topic.participated = is_ours || topic.participated;
+    topics.set(key, topic);
+    return true;
```

Comment:
I think this function could be a lot more readable
(1) `const topic` should be called `const topic_data` or similar; `const topic` sounds like the name not the dictionary.
(2) `Do we need the `topics.set()` at the end?  It looks like we're mutating-in-place.
(3) I think the `=== undefined` code path would be a lot better if it just did `topic = topics.set(key, {last_msg_id: -1, starred: new Set(), participated: false})` without returning, and then we just ran the same code as normal.  That way, we're not duplicating logic that would otherwise become confusing.

Reply:
Cool! Updated.